### PR TITLE
Expose ECDH shared key derivation

### PR DESCRIPTION
**Pull request by pavel-bc**
## Description

This PR exposes [ECDH Shared Key Computation](https://cryptobook.nakov.com/asymmetric-key-ciphers/ecdh-key-exchange) function on `PrivateKey` interface, which is commonly used across crypto applications. 

Internally it calls `ecdh_multiply()` from `TrezorCrypto`, computes SHA-256 hash from compressed version of resulting key - [in the same manner as libsecp256k1](https://crypto.stackexchange.com/a/57727)

## Testing instructions

Core library tests:

```sh
$ make -Cbuild
$ ./build/tests/tests tests --gtest_filter="PrivateKey*"
$ ./build/tests/tests tests --gtest_filter="TWPrivateKey*"
```

iOS tests:

```sh
$ ./tools/ios-test
```

Android tests:

```sh
$ ./tools/android-test
```

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/PrivateKey.cpp [hunk 1]
```diff
@@ -145,6 +146,25 @@ PublicKey PrivateKey::getPublicKey(TWPublicKeyType type) const {
     return PublicKey(result, type);
 }
 
+Data PrivateKey::getSharedKey(const PublicKey& pubKey, TWCurve curve) const {
+    if (curve != TWCurveSECP256k1) {
+        return {};
+    }
+
+    Data result;
+    result.resize(65);
```
Comment (catenocrypt) [inline src/PrivateKey.cpp:155]:
```suggestion
    Data result(PublicKey::secp256k1ExtendedSize);
```

File: src/PrivateKey.cpp [hunk 2]
```diff
@@ -145,6 +146,25 @@ PublicKey PrivateKey::getPublicKey(TWPublicKeyType type) const {
     return PublicKey(result, type);
 }
 
+Data PrivateKey::getSharedKey(const PublicKey& pubKey, TWCurve curve) const {
+    if (curve != TWCurveSECP256k1) {
+        return {};
```
Comment (catenocrypt) [inline src/PrivateKey.cpp:151]: For this error branch please add a negative unit test

File: src/PrivateKey.cpp [hunk 3]
```diff
@@ -145,6 +146,25 @@ PublicKey PrivateKey::getPublicKey(TWPublicKeyType type) const {
...
+    }
+
+    Data result;
+    result.resize(65);
+    bool success = ecdh_multiply(&secp256k1, bytes.data(),
+                                 pubKey.bytes.data(), result.data()) == 0;
+
+    if (!success) {
+        return {};
```
Comment (catenocrypt) [inline src/PrivateKey.cpp:160]: Trezor ecdh_multiply() never returns error, so error check here cannot be covered by test... :(   For the sake of coverage, I suggest reverting the if
```
    if (success) {
        ...
    }
```

**Review by catenocrypt — COMMENTED** [review]
Looks good.
Method should be exposed in  include/TrustWalletCore/TWPrivateKey.h
Then swift and kotlin tests can be added
Let me know if you need assistance

File: include/TrustWalletCore/TWPrivateKey.h [hunk 4]
```diff
@@ -60,6 +60,11 @@ struct TWPublicKey *_Nonnull TWPrivateKeyGetPublicKeyEd25519Extended(struct TWPr
 TW_EXPORT_METHOD
 struct TWPublicKey *_Nonnull TWPrivateKeyGetPublicKeyCurve25519(struct TWPrivateKey *_Nonnull pk);
 
+/// Computes an EC Diffie-Hellman secret in constant time
+/// Supported curves: secp256k1
+TW_EXPORT_METHOD
+TWData *_Nullable TWPrivateKeyGetSharedKey(struct TWPrivateKey *_Nonnull pk, struct TWPublicKey *_Nonnull publicKey, enum TWCurve curve);
```
Comment (catenocrypt) [inline include/TrustWalletCore/TWPrivateKey.h:66]: Add consts to const input parameters

**Comment by catenocrypt** [comment]
I would also request a test in `tests/interface/TWPrivateKeyTests.cpp` (too much overhead, I know). Watch out for wrapping returned value, so it is released.

File: include/TrustWalletCore/TWPrivateKey.h (hunk 4 above)
Comment (pavel-bc) [reply include/TrustWalletCore/TWPrivateKey.h]: @catenocrypt Are you sure about this? E.g. `sign` method below doesn't have this annotation in exposed wrapper. When I add it, it breaks iOS tests (`no symbol found for architecture` error)

Comment (pavel-bc) [reply include/TrustWalletCore/TWPrivateKey.h:66]: nevermind, seems like my mistake - forgot to add same qualifier to wrapper implementation

**Review by hewigovens — COMMENTED** [review]
lgtm, do you have  any libsecp256k1 test case to reference?

**Comment by pavel-bc** [comment]
@hewigovens Seems like [secp256k1](https://github.com/bitcoin-core/secp256k1/blob/master/src/modules/ecdh/tests_impl.h) doesn't have any explicit test vectors.

I've added 1st valid test vector from [Project Wycherproof](https://github.com/google/wycheproof/blob/master/testvectors/ecdh_secp256k1_test.json#L31) (which is used by [trezor-crypto](https://github.com/trustwallet/wallet-core/blob/master/trezor-crypto/crypto/tests/test_wycheproof.py)), however it uses it's own encoding formats & only [validates x coordinate](https://crypto.stackexchange.com/a/57727/87769) - so I had to manually verify test results using several tools:

- [ASN.1 Decoder](https://holtstrom.com/michael/tools/asn1decoder.php)
- [Bitcoin Key Compression Tool](https://iancoleman.io/bitcoin-key-compression/)
- [Online Hash Value Calculator](https://www.fileformat.info/tool/hash.htm) - check `Binary Hash`

Please check the comments in `PrivateKeyTests.cpp`

**Review by KaycoinsInc — APPROVED** [review]

**Comment by KaycoinsInc** [comment]
> ## Description
> 
> This PR exposes [ECDH Shared Key Computation](https://cryptobook.nakov.com/asymmetric-key-ciphers/ecdh-key-exchange) function on `PrivateKey` interface, which is commonly used across crypto applications. 
> 
> Internally it calls `ecdh_multiply()` from `TrezorCrypto`, computes SHA-256 hash from compressed version of resulting key - [in the same manner as libsecp256k1](https://crypto.stackexchange.com/a/57727)
> 
> ## Testing instructions
> 
> Core library tests:
> 
> ```sh
> $ make -Cbuild
> $ ./build/tests/tests tests --gtest_filter="PrivateKey*"
> $ ./build/tests/tests tests --gtest_filter="TWPrivateKey*"
> ```
> 
> iOS tests:
> 
> ```sh
> $ ./tools/ios-test
> ```
> 
> Android tests:
> 
> ```sh
> $ ./tools/android-test
> ```
> 
> ## Types of changes
> 
> 
> 
> 
> 
> * New feature (non-breaking change which adds functionality)
> 
> 
> 
> ## Checklist
> 
> 
> 
> - [x] Prefix PR title with `[WIP]` if necessary.
> - [x] Add tests to cover changes as needed.
> - [x] Update documentation as needed.
> - [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
> - [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).
> 

Very nice job.

**Comment by pavel-bc** [comment]
@hewigovens @KaycoinsInc @catenocrypt Could you guys please help with Linux CI? Haven't touched anything CMake-related in last two commits

**Review by hewigovens — APPROVED** [review]

**Comment by hewigovens** [comment]
linux CI boost version might changed

**Comment by hewigovens** [comment]
https://github.com/actions/virtual-environments/issues/2667

**Comment by hewigovens** [comment]
will fix it in this pr: https://github.com/trustwallet/wallet-core/pull/1327

**Comment by hewigovens** [comment]
@pavel-bc can you rebase master? CI should be fixed

**Review by catenocrypt — APPROVED** [review]

**Pull request merged**